core/shared/src/main/scala/magnolia.scala

```diff
@@ -706,15 +706,17 @@ private[magnolia] object CompileTimeState {
   object Stack {
     // Cheating to satisfy Singleton bound (which improves type inference).
     private val dummyContext: whitebox.Context = null
-    private val global = new Stack[dummyContext.type]
-    private val workSet = mutable.Set.empty[whitebox.Context#Symbol]
+    private val threadLocalStack = ThreadLocal.withInitial[Stack[dummyContext.type]](() => new Stack[dummyContext.type])
+    private val threadLocalWorkSet = ThreadLocal.withInitial[mutable.Set[whitebox.Context#Symbol]](() => mutable.Set.empty)
```

Comment:
I haven't looked into magnolia internals nearly enough, so I have zero context - forgive me if it's a stupid question 😄 

I assume no merging is required for different instances of this `ThreadLocal` because everything from the same compilation unit should end up in the same stack, right?

Reply:
That would be my assumption but it really depends on Hydra.
I would not expect it to switch threads for a single macro expandsion.
But on the other hand if it's using `Future` this won't help.